plugins/modules/edpm_nova_derive_pci_passthrough_devicespec.py

```diff
@@ -0,0 +1,649 @@
+                user_config['address'] = get_pci_addr_from_ifname(user_config['devname'])
+                del user_config['devname']
+            else:
+                user_config_copy.append(user_config)
+                continue
+        if ('address' in user_config
+                or 'product_id' in user_config):
+            passthrough_tmp, del_user_config = get_passthrough_config_for_all_pf(
+                user_config, system_configs, allocated_pci)
+
+            """ If del_user_config is set, do not add to derived_config or
+                user_config_copy
+            """
+            if not del_user_config:
+                if passthrough_tmp:
+                    derived_config.extend(passthrough_tmp)
+                else:
+                    user_config_copy.append(user_config)
+        else:
+            user_config_copy.append(user_config)
+    return (user_config_copy, derived_config)
+
+
+def get_nova_userconfigs(nova_conf_file):
+    configs = []
+    src_config = ""
+    # Currently only "device_spec" is valid
+    valid_spec = ('device_spec')
+
+    # Read user config from edpm_nova_config conf file
+    try:
+        with open(nova_conf_file) as nova_config:
+            for line in nova_config:
+                pf = line.strip()
+                pf = line.replace('\'', '\"')
+                if pf.startswith(valid_spec):
+                    # make a copy for replace after derive
+                    src_config += line
+                    obj = pf.split('=')[1]
+                    spec = json.loads(obj)
+                    configs.append(spec)
+    except IOError:
+        configs = []
+    return (configs, src_config)
+
+
+def _run_derive_pci(sriov_nova_conf_file,
+                    sriov_phydev_map=None):
+    system_configs = get_sriov_configs()
+    user_configs, input_config = get_nova_userconfigs(sriov_nova_conf_file)
+    lines = ""
+
+    # Check if user config list is valid
+    if not isinstance(user_configs, list):
+        raise InvalidConfigException('user_config specified is not a list {!r}'.format(user_configs))
+
+    user_config_copy, derived = generate_combined_configuration(
+        user_configs, system_configs, sriov_phydev_map)
+
+    """ If the derivation does not bring in any changes, the user_config_copy list
+        and user_configs shall be same. The pci_passthrough_devspec
+        shall be updated only when there is a change needed due to NIC Partition
+    """
+    if user_config_copy != user_configs:
+        for item in (user_config_copy + derived):
+            line = _PASSTHROUGH_DEVSPEC_KEY + '=' + json.dumps(item)
+            lines += "".join(line) + '\n'
+        # Replace the new derived devspec into conf
+        with open(sriov_nova_conf_file) as nova_config:
+            file_data = nova_config.read()
+            update = file_data.replace(input_config, lines)
+        with open(sriov_nova_conf_file, 'w') as nova_config:
+            nova_config.write(update)
```

Comment:
this script should not be using a nova config files as input but even if we ignore that its not ok for this script to modify that.